lib/ip_access_control.ex

```diff
@@ -184,18 +185,18 @@ defmodule IpAccessControl do
   end
 
   def allowed?(remote_ip, allow_list) when is_binary(remote_ip) do
-    case :inet.parse_strict_address(to_charlist(remote_ip)) do
+    case BitwiseIp.parse(remote_ip) do
```

Comment:
The distinction between `parse` and `encode` feels a little artificial to me. I know what and why, but ideally, I could simply do something intelligent here.

Also, I know I’m inviting trouble with the `is_tuple` rather than `is_tuple(remote_ip) and (tuple_size(remote_ip) == 4 or tuple_size(remote_ip) == 6`, but that also feels iffy to me.

Reply:
> The distinction between `parse` and `encode` feels a little artificial to me. I know what and why, but ideally, I could simply do something intelligent here.

Mmm. Not sure I want to change that, though. `parse` is to `to_string` as `encode` is to `decode`; they're doing distinct things with distinct types. I think it's clearer to have a little extra code if you want to be super flexible with the types you're accepting. Unless maybe there's a good "type casting" function name. 🤔 `BitwiseIp.from/1`, perhaps?
```elixir
BitwiseIp.from({127, 0, 0, 1}) == BitwiseIp.from("127.0.0.1")
```

> Also, I know I’m inviting trouble with the `is_tuple` rather than `is_tuple(remote_ip) and (tuple_size(remote_ip) == 4 or tuple_size(remote_ip) == 6`, but that also feels iffy to me.

Yeah, I kind of got that way with the `BitwiseIp` type specs. I could declare a lot stronger type information, like using `0..0xFFFFFFFF` instead of `integer()` for IPv4-encoded ints. In the end, it's about striking a balance. Elixir is a very dynamically typed language; it generally trusts people to avoid shooting their own feet. I definitely feel there are diminishing returns to worrying about it too much. Otherwise you'd start having copy-pasta function guards like
```elixir
def allowed?({a, b, c, d}, allow_list) when a in 0..255 and b in 0..255 and c in 0..255 and d in 0..255 and is_list(allow_list) and ... do
```
I guess the term that comes to mind is YAGNI. (Kind of applies to the binary encoding of IP addresses, truth be told. 😅 But I enjoy being a crotchety grandpa about kids these days not treating IPs like the unsigned ints that they are. Grumble grumble.)

Reply:
`BitwiseIp.from` _might_ be a good way to do this as a straight-up usability thing. It’s not a big deal as the code that I have _works_. The thought here is pushing some of the complexity and work _down_ so that the API is cleaner and easier to work with.

If you choose not to unify them in some way (and it’s _perfectly fine not to do so_), it’s probably worth describing the different utilities in the documentation as you write it up.